### PR TITLE
Table component with filter and search

### LINKED ISSUE #569: Edit Style of Table Component
## 🗒 General 

It would be amazing if someone could update the [sortable table component ](http://storybook.backstage.io/?path=/story/sortable-table--default) styles!

## 💻 Usage

The sortable table is a robust way to view a lot of categorized information at once. By giving users the power to modify their views of the inputs, they can comprehend the data in a much easier way. 

### Default state
Default

### Hover
Hover

### Active
Focus

## 📐 Specs

Specs

### Rules
1. Text input columns are always left aligned for easier read. The sort arrow icon will go to the right of the column header.  _Caveat: this can be adjusted for different languages._
2. Numerical input columns are always right aligned for easier read. The sort arrow icon will go to the left of the column header.

### Example
Screenshot 2020-04-16 at 4 38 25 PM

## 🔮 Future

- Status
- Additional types of input

### PR DESCRIPTION
**Pull request by hooloovooo**
## Hey, I just made a Pull Request!

Making a PR for feedback on the approach so far. I feel using react-table is a very flexible solution that we can keep iterating on as new requirements pop up.

#### :heavy_check_mark: Checklist

- [x] All tests are passing `yarn test`
- [x] Screenshots attached (for UI changes)
- [ ] Relevant documentation updated
- [x] Prettier run on changed files
- [ ] Tests added for new functionality
- [ ] Regression tests added for bug fixes

2020-04-29-160716_1453x675_scrot

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -37,8 +37,13 @@
...
   },
   "dependencies": {
+    "@backstage/core": "^0.1.1-alpha.4",
     "@types/classnames": "^2.2.9",
-    "moment": "^2.24.0"
+    "@types/react-table": "^7.0.16",
+    "moment": "^2.24.0",
+    "react-table": "^7.0.4",
+    "react-virtualized-auto-sizer": "^1.0.2",
+    "react-window": "^1.8.5"
```
Comment (Rugvip) [inline package.json:46]: none of these here please, should all go in the package using them

Comment (hooloovooo) [reply package.json:46]: Woops, sorry. Though they would be added to the correct package automatically when doing yarn add -W <package>. Guess not.

**Comment by stefanalund** [comment]
Please wrap the stories in InfoCards. Styling is not correct either but maybe that’s work for another PR? @katz95

**Comment by stefanalund** [comment]
Another note: when this is done the example plugin you get when running the scaffolder should use this Table.

**Comment by hooloovooo** [comment]
> Please wrap the stories in InfoCards. Styling is not correct either but maybe that’s work for another PR? @katz95

I will talk to @katz95 about styling, but i'm mainly looking for feedback on the technical solution and API for now. Currently the API is very minimal, but should allow for expansion with virtualized tables for example.

File: packages/core/src/components/Table/Table.tsx [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+  );
+};
+
+type TableProps = {
+  data: Array<any>;
+  columns: Array<any>;
+};
+
+const Table: FC<TableProps> = ({ columns, data }) => {
+  0;
```
Comment (stefanalund) [inline packages/core/src/components/Table/Table.tsx:68]: Maybe its my TS foo that isn't strong enough, but this looks strange to me.

File: packages/core/src/components/Table/Table.tsx [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+    },
+    [prepareRow, rows],
+  );
+
+  const renderTableColumn = (type, column) => {
+    switch (type) {
+      default:
+        return (
+          <TableCell
+            align={type === 'numeric' ? 'right' : 'left'}
```
Comment (stefanalund) [inline packages/core/src/components/Table/Table.tsx:112]: Feels like alignment will be a relevant property for non-numeric values as well. One common example is statuses such as `<StatusOK>`.

File: packages/core/src/components/Table/Table.tsx [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+const GlobalFilter: FC<GlobalFilterProps> = ({
+  setGlobalFilter,
+  preGlobalFilteredRows,
+  globalFilter,
+}) => {
+  const count = preGlobalFilteredRows.length;
+
+  return (
+    <span>
```
Comment (stefanalund) [inline packages/core/src/components/Table/Table.tsx:48]: Could this be `<>`?

File: packages/core/src/components/Table/Table.tsx [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+          setGlobalFilter(e.target.value || undefined); // Set undefined to remove the filter entirely
+        }}
+      />
+    </span>
+  );
+};
+
+type TableProps = {
+  data: Array<any>;
+  columns: Array<any>;
```
Comment (stefanalund) [inline packages/core/src/components/Table/Table.tsx:64]: A toggle for showing/hiding the search bar is relevant I think.

File: packages/core/src/components/Table/Table.tsx (hunk 2 above)
Comment (Rugvip) [reply packages/core/src/components/Table/Table.tsx:68]: It looks strange to me too :grin:

File: packages/core/src/components/Table/Table.tsx (hunk 4 above)
Comment (Rugvip) [reply packages/core/src/components/Table/Table.tsx:48]: Or just removed

File: packages/core/src/components/Table/Table.tsx (hunk 5 above)
Comment (Rugvip) [reply packages/core/src/components/Table/Table.tsx:64]: In general I'm wondering if we really want to create out own API here. Internally I think we solve 80% of the usages by just copying stuff, but the docs have been important for the rest imo. Compare to the docs at https://material-table.com/ for example, not sure we want to replicate that.

The strategy could be that we provide this table component to solve 80%, and then let people use `react-table` directly for the rest.

Comment (stefanalund) [reply packages/core/src/components/Table/Table.tsx:64]: @Rugvip my take is that we should have a super simple table component that covers 60-80%. Our internal version has grown to cover ~95% which also meant more complexity and poor docs.

One way to make this clear is to have 2 different components in our Storybook. One for the simple use-cases (this one) and a more powerful (Complex Table) version that uses Material Table as is.

Comment (hooloovooo) [reply packages/core/src/components/Table/Table.tsx:64]: I was looking at material-table, but could not find any support for virtualization. If that is a requirement at some point we might have to find a third table as well. That why i went with react-table for this example as it allows us to start small and reasonably easy add the features we want without adding stuff we don't need.
Other than that material-table looks really nice though, so if i'm mistaken about it supporting virtualization or if we know we won't need it i'm all for it. :)

File: packages/core/src/components/Table/Table.stories.tsx [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    sortType: 'datetime',
+    Cell: ({ value }) => value.toLocaleDateString(),
+  },
+];
+
+const testData100 = generateTestData(100);
+
+export const DefaultTable = () => {
+  const columns = React.useMemo(() => testColumns, []);
+  const data = React.useMemo(() => testData100, []);
```
Comment (freben) [inline packages/core/src/components/Table/Table.stories.tsx:66]: why are these memoized?

File: packages/core/src/components/Table/Table.tsx [hunk 7]
```diff
@@ -0,0 +1,177 @@
...
+import TableBody from '@material-ui/core/TableBody';
+import TableCell from '@material-ui/core/TableCell';
+import TableContainer from '@material-ui/core/TableContainer';
+import TableHead from '@material-ui/core/TableHead';
+import TableRow from '@material-ui/core/TableRow';
+import TableSortLabel from '@material-ui/core/TableSortLabel';
+
+import TextField from '@material-ui/core/TextField';
+
+//const useStyles = makeStyles<typeof BackstageTheme>(theme => ({ }));
```
Comment (freben) [inline packages/core/src/components/Table/Table.tsx:33]: remove when wip is done

File: packages/core/src/components/Table/Table.tsx [hunk 8]
```diff
@@ -0,0 +1,177 @@
...
+  globalFilter,
+}) => {
+  const count = preGlobalFilteredRows.length;
+
+  return (
+    <span>
+      <TextField
+        value={globalFilter || ''}
+        placeholder={`${count} records...`}
+        id="standard-basic"
```
Comment (freben) [inline packages/core/src/components/Table/Table.tsx:52]: What is this id for? There may be more than one table per page, best not to give them IDs.

File: packages/core/src/components/Table/Table.tsx [hunk 9]
```diff
@@ -0,0 +1,177 @@
...
+  } = useTable(
+    {
+      columns,
+      data,
+    },
+    useGlobalFilter,
+    useSortBy,
+  );
+
+  //Used for react-window rendering. Which means not at all right now.
```
Comment (freben) [inline packages/core/src/components/Table/Table.tsx:87]: remember to have space after `//`, maybe you missed enabling linting or prettier?

File: packages/core/src/components/Table/Table.tsx (hunk 5 above)
Comment (freben) [reply packages/core/src/components/Table/Table.tsx:64]: I agree with having a prop to hide the search. A great feature set for a basic table would be sorting by clicking headers, and filtering that just works. Having both by default and then have props for (1) disabling sorting, or (2) hiding headers entirely, and (3) hiding the filter box, would go a long way toward a great but easy to use table for that 80% case.

File: packages/core/src/components/Table/Table.tsx [hunk 10]
```diff
@@ -0,0 +1,177 @@
...
+  const renderRow = useCallback(
+    ({ index, style }) => {
+      const row = rows[index];
+      prepareRow(row);
+      return (
+        <TableRow style={style} {...row.getRowProps()}>
+          {row.cells.map(cell => {
+            return (
+              <TableCell {...cell.getCellProps()}>
```
Comment (freben) [inline packages/core/src/components/Table/Table.tsx:96]: do the generated props deal with `key` as well?

Comment (hooloovooo) [reply packages/core/src/components/Table/Table.tsx:91]: > do the generated props deal with `key` as well?

Yup, that's taken care of by row/cell.getRow/CellProps()

**Comment by hooloovooo** [comment]
Had to wrestle with typescript a bit to get the typing to work, and i still get an error from prettier saying
`[error] packages/core/src/components/Table/Table.tsx: SyntaxError: '=' expected. (19:13)
[error]   17 | import React, { FC } from 'react';
[error]   18 | import { useTable, useGlobalFilter, useSortBy } from 'react-table';
[error] > 19 | import type { Column } from 'react-table';
[error]      |             ^
[error]   20 |
[error]   21 | import MUITable from '@material-ui/core/Table';
[error]   22 | import TableBody from '@material-ui/core/TableBody';
[error] packages/core/src/components/Table/react-table-config.d.ts: SyntaxError: '=' expected. (51:13)
[error]   49 | } from 'react-table'
[error]   50 |
[error] > 51 | import type { ColumnProps } from './Table';
[error]      |             ^
[error]   52 |
[error]   53 | declare module 'react-table' {
[error]   54 |   // take this file as-is, or comment out the sections that don't apply to your plugin configuration
`

Does anyone who is stronger with the typescript force know what i'm doing wrong here?

File: packages/core/src/components/MaterialTable/MaterialTable.stories.tsx [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+  }
+
+  return data;
+};
+
+const testColumns = [
+  {
+    title: 'Column 1',
+    field: 'col1',
+    headerStyle: {
```
Comment (stefanalund) [inline packages/core/src/components/MaterialTable/MaterialTable.stories.tsx:45]: This is ideally not something we should expose (I think -- @katz95?).

Comment (Rugvip) [reply packages/core/src/components/MaterialTable/MaterialTable.stories.tsx:45]: +1, at least not encourage it imo

Comment (hooloovooo) [reply packages/core/src/components/MaterialTable/MaterialTable.stories.tsx:45]: +1. It's mostly a easy way to show how to do this in material-table. I guess we could create a layer between the headers sent in the our component and what's passed down to material-table. Might be some other way, but i couldn't find it.

Comment (stefanalund) [reply packages/core/src/components/MaterialTable/MaterialTable.stories.tsx:45]: Should at least not show it in the examples :)

**Comment by stefanalund** [comment]
Fixes https://github.com/spotify/backstage/issues/569

**Comment by stefanalund** [comment]
Could you replace the screenshot with a more recent one?

**Comment by hooloovooo** [comment]
Seems like some hook ran prettier on previous commits when i merged master. 
🤷‍♂️

**Comment by stefanalund** [comment]
@hooloovooo yarn lock conflicts :)

**Review by stefanalund — APPROVED** [review]

**Pull request merged**